### PR TITLE
allow constant propagation for detach

### PR DESCRIPTION
**Pull request by laithsakka**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #172589

when some libraries are imported during pre-grad passes, for one of vllm models, we end up executing 
some global code that looks like
```
def inv_softplus(bias):
    if not isinstance(bias, torch.Tensor):
        bias = torch.tensor(bias)  # Creates tensor from scalar
    out = bias.expm1().clamp_min(1e-6).log()
    if out.numel() == 1:
        return out.item()  # Fails with DataDependentOutputException!
    return out
x = {inv_softplus(1.0)}
```
This is end up throwing DataDependentOutputException.

This could can be avoided if constant propagation is not interrupted.  torch.tensor(scalar) internally calls 
detach_() (I do not know why the detach happens here , but because it happens constant propagation 
is not happening on "1.0" since detach_ has the inplace_view tag and Inside
FakeTensorMode.__torch_dispatch__ ops with inplace_view tag skip the constant propagation block

detach does not change constant values so special handling it here. 
```
and (
    torch.Tag.inplace_view not in func.tags
    or func is aten.detach_.default  # Allow detach_ through
)
```

**Why detach_ is safe to special-case**
We dispatch size/stride/numel on the FakeTensor not its constant, so bail on inplace_view.
 Example: fake_a.transpose_(0,1) would mutate fake_a.constant in-place, changing its
shape from (2,3) to (3,2), while fake_a.shape still reports (2,3) → divergence.
However, detach_ is safe: it only mutates requires_grad (not shape/stride/data),
and constants are used purely for their values, not autograd.

### CODE REVIEW

**Review by eellison — APPROVED** [review]

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -2524,7 +2524,12 @@ def _dispatch_impl(
         if (
             isinstance(func, torch._ops.OpOverload)
             and torch.Tag.nondeterministic_seeded not in func.tags
-            and torch.Tag.inplace_view not in func.tags
+            # detach_ is a view-like operation that doesn't modify tensor data,
+            # it just detaches from the autograd graph. We should preserve constants
+            # for detach_ since the underlying data is unchanged.
+            and (
+                torch.Tag.inplace_view not in func.tags or func is aten.detach_.default
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:2532]: hmm... I could believe this change is reasonable but I'm not sure that I agree with this justification:

> Unlike other inplace_view operations (e.g., transpose_, squeeze_) which change tensor metadata (shape, stride, storage offset), detach_ only detaches the tensor from the autograd graph. The underlying data and metadata remain unchanged, so preserving the constant backing is safe and correct.

the autograd metadata counts as "metadata" on the tensor. So in the same way that `x.t_()` inplace-mutates the metadata on `x` (the sizes/strides), `x.detach_()` will also inplace-mutate the metadata on `x` (the `grad_fn` field).

Are we saying it's ok to do that for constant prop? I need to refresh my memory a bit on what the invariants around constant prop are (and why we ban inplace views in the first place)

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -2524,7 +2524,12 @@ def _dispatch_impl(
...
+            )
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:2533]: also - is it possible to add a test?

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (laithsakka) [reply torch/_subclasses/fake_tensor.py:2532]: it wont effect the propagated constant value basically

Comment (eellison) [reply torch/_subclasses/fake_tensor.py:2532]: @bdhirsh, torch dispatch subclasses like fake tensor are below the autograd metadata, as far as i understand.

File: torch/_subclasses/fake_tensor.py (hunk 2 above)
Comment (laithsakka) [reply torch/_subclasses/fake_tensor.py:2533]: i tried hard could not i can spend 1 more hour i have a vllm repo lol

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_subclasses/fake_tensor.py:2532]: so my mental model is something like:

(1) since __torch_dispatch__ runs below autograd, then when you call an op that produces a fresh `Tensor` object, the autograd engine is waiting for you to return that tensor before populating the autograd metadata onto it (e.g. `grad_fn`). This makes it safe to e.g. perform non-differentiable operations inside of __torch_dispatch__, and assume that autograd will populate autograd metadata on the output for you properly.

(2) that only really applies to functional ops though? For metadata mutations, if the input FakeTensor has some autograd metadata, that metadata is still "visible" (both inside of the __torch_dispatch__ call, and also visible to the user before/after)

Comment (bdhirsh) [reply torch/_subclasses/fake_tensor.py:2532]: i don't want to block this change though if it fixes concrete problem though (I just don't quite remember the use case around fake tensor constant prop and how it relates to this vLLM failure)

**Review by eellison — DISMISSED** [review]
lets wait for @bdhirsh  or @ezyang to approve, not sure i have the right context on torch dispatch autograd meta

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (laithsakka) [reply torch/_subclasses/fake_tensor.py:2532]: yeh i dont know i will leave it for experts but the general context
in pre gard passed we do some import-> go through import fbcode -> import bunch of 3rd party libraries
all that happen with fake mode enabled !

then we end up doing 
```
def inv_softplus(bias):
    if not isinstance(bias, torch.Tensor):
        bias = torch.tensor(bias)  # Creates tensor from scalar
    out = bias.expm1().clamp_min(1e-6).log()
    if out.numel() == 1:
        return out.item()  # Fails with DataDependentOutputException!
    return out
x = {inv_softplus(1.0)}
```
which hit a DDE. 
because shape_env is null? why shape_env is null.

the whole initializations is wrapped with :
with torch._guards.tracing(None), unset_fake_temporarily(), FakeTensorMode():
in this function
@init_once_fakemode

see full context in this post:
https://fb.workplace.com/groups/1075192433118967/permalink/1846231192681750/

one way to fix it is to remove the code from running globally but this can hit anytime again !

**Comment by laithsakka** [comment]
@ezyang idk if you are familiar with this code, I am postive but not confident about the change.

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (bdhirsh) [reply torch/_subclasses/fake_tensor.py:2532]: just to cover some other options: if we are importing 3rd party libs *inside* of the pregrad passes, and those 3rd party libs can execute arbitrary python code (including... compute?), then why don't we wrap the imports with a `temporarily_disable_fake_mode`? (I forget the exact name of the API). That should reduce pressure on pre-grad passes requiring FakeMode to handle running arbitrary user code.

Comment (bdhirsh) [reply torch/_subclasses/fake_tensor.py:2532]: Oh i missed what you mentioned at the bottom - it sounds like we are running the init code under `unset_fake_temporarily()` already? (if so, why are the imports running with a fake mode active?)

Comment (laithsakka) [reply torch/_subclasses/fake_tensor.py:2532]: we do unset the fake tensor mode and then we set a new fake tensor mode
```

def init_once_fakemode(fn: Callable[..., Any]) -> Callable[[], Any]:
    """Wrapper around lazy init functions in fx_passes/"""

    @functools.cache
    @functools.wraps(fn)
    def lazy_init() -> Any:
        counters_ref = counters[backend].copy()

        with torch._guards.tracing(None), unset_fake_temporarily(), FakeTensorMode():
            result = fn()

        # clear view matches encountered during tracing
        counters[backend] = counters_ref

        return result

    return lazy_init
```

this wraps
```
@init_once_fakemode
def lazy_init():
    from . import (  # noqa: F401  # noqa: F401
        apply_gumbel_max_trick,
        efficient_conv_bn_eval,
        split_cat,
    )

    if config.is_fbcode():
        from . import fb  # type: ignore[attr-defined]  # noqa: F401
```
which is called here:
```
def pre_grad_passes(
    gm: torch.fx.GraphModule,
    example_inputs: Sequence[object] = (),
    add_passes: str | None = None,
    remove_passes: str | None = None,
) -> torch.fx.GraphModule:
    """
    Apply passes on the input FX graph using Torch IR.

    WARNING:
    The IR before grad is not functional or normalized, so it is harder
    to write passes on this IR.  Passes must be safe with respect to
    aliasing and mutation and need to handle all possible arg schemas.

    Consider adding a new pass to post_grad.py or joint_graph.py which
    are after functionalization and normalization.
    """
    if config.pattern_matcher:
        lazy_init()
        if hasattr(
        ....
```

File: torch/_subclasses/fake_tensor.py (hunk 2 above)
Comment (laithsakka) [reply torch/_subclasses/fake_tensor.py:2533]: could not get a detach that is not optimized away/ ~~

Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:2533]: @laithsakka did you try asking Claude to make the repro? If you give it the test case failing in situ in vllm it does a pretty good job at it.

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:2532]: I think the thing I am worried about here is not that const propping through `detach_` is wrong, I'm worried about the *inplace mutation* causing potential problems.

**Review by ezyang — COMMENTED** [review]
.

**Comment by laithsakka** [comment]
> @claude

@claude can you suggest a good informative comment to add to the change that justifies why its ok

**Review by ezyang — APPROVED** [review]
I'm preemptively approving this pending a comment
